presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java

```diff
@@ -0,0 +1,214 @@
+            this.parameters = requireNonNull(parameters, "parameters is null");
+            this.queryExplainer = requireNonNull(queryExplainer, "queryExplainer is null");
+        }
+
+        @Override
+        protected Node visitShowStats(ShowStats node, Void context)
+        {
+            checkArgument(node.getQuery().getQueryBody() instanceof QuerySpecification);
+            checkState(queryExplainer.isPresent(), "Query explainer must be provided for SHOW STATS SELECT");
+            QuerySpecification specification = (QuerySpecification) node.getQuery().getQueryBody();
+
+            checkArgument(specification.getFrom().isPresent());
+            checkArgument(specification.getFrom().get() instanceof Table);
+            Table table = (Table) specification.getFrom().get();
+            TableHandle tableHandle = getTableHandle(node, table.getName());
+            Constraint<ColumnHandle> constraint = getConstraint(node, tableHandle);
+
+            TableStatistics tableStatistics = metadata.getTableStatistics(session, tableHandle, constraint);
+
+            List<String> statisticsNames = findUniqueStatisticsNames(tableStatistics);
+
+            List<String> resultColumnNames = buildColumnsNames(statisticsNames);
+            SelectItem[] selectItems = buildSelectItems(resultColumnNames);
+
+            Map<ColumnHandle, String> columnNames = getStatisticsColumnNames(tableStatistics, node, table.getName());
+            List<Expression> resultRows = buildStatisticsRows(tableStatistics, columnNames, statisticsNames);
+
+            return simpleQuery(selectList(selectItems),
+                    aliased(new Values(resultRows),
+                            "table_stats_for_" + table.getName(),
+                            resultColumnNames));
+        }
+
+        @Override
+        protected Node visitNode(Node node, Void context)
+        {
+            return node;
+        }
+
+        private Constraint<ColumnHandle> getConstraint(ShowStats node, TableHandle tableHandle)
+        {
+            checkArgument(node.getQuery().getQueryBody() instanceof QuerySpecification);
+            QuerySpecification specification = (QuerySpecification) node.getQuery().getQueryBody();
+
+            if (specification.getWhere().isPresent()) {
+                return getConstraintWithWhere(node);
+            }
+            return Constraint.alwaysTrue();
+        }
+
+        private Constraint<ColumnHandle> getConstraintWithWhere(ShowStats node)
+        {
+            checkArgument(node.getQuery().getQueryBody() instanceof QuerySpecification);
+            QuerySpecification specification = (QuerySpecification) node.getQuery().getQueryBody();
+
+            Plan plan = queryExplainer.get().getLogicalPlan(session, new Query(Optional.empty(), specification, Optional.empty(), Optional.empty()), parameters);
+            Map<Symbol, Type> types = plan.getTypes();
+
+            Optional<TableScanNode> scanNode = findTableScanNode(plan);
```

Comment:
Why do we need to do this (find table scan, filter, etc)? The TableScan node should already have all the necessary information for the connector to provide the stats for the table scan with pushed-down predicates.